graftm/tree_decorator.py

```diff
@@ -1,309 +1,187 @@
-            for tip in self.tree.tips():
-                ancestors = tip.ancestors()
-                
-                ancestor_tax = []
-                for ancestor in ancestors:
-                    if ancestor.name:
-                        split_node_name = ancestor.name.split(':')
-                        if len(split_node_name) == 2:
-                            ancestor_tax.append(split_node_name[1])
-                        elif len(split_node_name) == 1:
-                            pass
-                        else:
-                            raise Exception("Malformed node name: %s" % ancestor.name)
-                        
-                tax_list = list(reversed(ancestor_tax))
-                tax_name = tip.name.replace(" ", "_")
-                if len(tax_list) < 1:
-                    logging.debug("Species %s not decorated within the tree, using provided tax string as substitute" % (tax_name))
-                    if tax_name in self.taxonomy:
-                        tax_string = '; '.join(self.taxonomy[tax_name])
-                    else:
-                        logging.warning("No taxonomy found for species %s!" % (tax_name))
-                        tax_string = "Unknown"
-                else:
-                    tax_string = '; '.join(tax_list)
-                output_line = "%s\t%s\n" % (tax_name, tax_string)
-                out.write(output_line)
-    
-    def _get_tax_index(self, ancestry):
+        taxonomy: TaxonomyObject
+                 
+        Returns
+        -------
+        skbio TreeNode object
         '''
-        Iterate through ancestor nodes to find the current taxonomy index (i.e.
-        the rank that should be assigned to the current node.
+        self.max_taxonomy_index=taxonomy.max
+        consistency_check = 1
+        seen_taxonomy_rank = {}
+        to_decorate = []
         
-        Inputs:
-            ancestry: list
-                TreeNode.ancestory list that contains all nodes above the 
-                current.
-        Outputs:
-            Current index of taxonomy. Returns None if node has no ancestors 
-            (i.e. is root)
-                
-        '''
-        current_index = 0
-        ancetor_rank_number = 0
-        ancestry = [x for x in ancestry if not x.is_root()]
-        if any(ancestry):
-            ancestor_tax_list = []
-            for ancestor in ancestry:
-                ai = self.encountered_nodes[ancestor]
-                
-                if ancestor.name:
-                    split_node_name = ancestor.name.split(':')
-                    if len(split_node_name) == 2:
-                        ancestor_tax_list.append(split_node_name[1])
-                    elif len(split_node_name) == 1:
-                        pass
+        for node in self.tree.preorder():    
+            tax_string_array = []
+            for rank in range(0, self.max_taxonomy_index):
+                node_classifications = []
```

Comment:
I'm not sure what difference that would make in this instance sorry.

I've implemented a workaround so sets are not used at all. They were kinda used in a dirty way here anyways.